MapboxNavigation/NavigationMapView.swift

```diff
@@ -1335,3 +1365,129 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
+        highlightedBuildingsLayer?.fillExtrusionHeightTransition = MGLTransition(duration: 0.8, delay: 0)
+        highlightedBuildingsLayer?.fillExtrusionOpacityTransition = MGLTransition(duration: 0.8, delay: 0)
+        
+        if let highlightedBuildingLayer = highlightedBuildingsLayer {
+            style?.addLayer(highlightedBuildingLayer)
+        }
+        
+        return highlightedBuildingsLayer
+    }
+    
+    private func removeBuildingsLayer() {
+        guard let highlightedBuildingsLayer = highlightedBuildingsLayer else { return }
+        
+        style?.removeLayer(highlightedBuildingsLayer)
+        self.highlightedBuildingsLayer = nil
+    }
+
+    private func buildingId(for coordinate: CLLocationCoordinate2D) -> Int64 {
+        let screenCoordinate = convert(coordinate, toPointTo: self)
+
+        // To increase a chance of selecting correct building ID filter out
+        // features which do not contain appropriate attributes.
+        let features = visibleFeatures(at: screenCoordinate).filter {
+            let extrude = $0.attribute(forKey: "extrude") as? String
+            let type = $0.attribute(forKey: "type") as? String
+            return extrude == "true" && type == "building"
+        }
+        
+        if let feature = features.first, let buildingId = feature.identifier as? Int64 {
+            return buildingId
+        }
+        
+        return BuildingHighlightAttributes.kDefaultIdentifier
+    }
+    
+    private func highlightBuildings(_ buildings: [BuildingHighlightAttributes], in3D: Bool = false, extrudeAll: Bool = false) {
+        // In case if array with highlighted building attributes is empty - do nothing.
+        if buildings.isEmpty { return }
+        // Add layer which will be used to highlight buildings if it wasn't added yet.
+        guard let highlightedBuildingsLayer = addBuildingsLayer() else { return }
+        
+        if extrudeAll {
+            highlightedBuildingsLayer.predicate = NSPredicate(format: "extrude = 'true' AND type = 'building' AND underground = 'false'")
+        } else {
+            // Form a predicate to filter out the other buildings from the datasource so only the desired ones are included.
+            highlightedBuildingsLayer.predicate = NSPredicate(format: "extrude = 'true' AND type = 'building' AND underground = 'false' AND $featureIdentifier IN %@", buildings.map { $0.identifier })
+        }
+        
+        // Buildings with IDs will be highlighted with provided color. Rest of the buildings will be highlighted, but kept at a uniform color.
+        let highlightedBuildingHeightExpression = "MGL_MATCH($featureIdentifier, \(buildings.map { "\($0.identifier), \(in3D ? "height" : "0"), " }.joined(separator: ""))\(extrudeAll ? "height" : "0"))"
+        let highlightedBuildingColorExpression = "MGL_MATCH($featureIdentifier, \(buildings.map { "\($0.identifier), %@, " }.joined(separator: ""))%@)"
```

Comment:
Avoid using concatenation or string interpolation to build an expression format string (or any kind of format string, for that matter). If the style happened to have a layer whose identifiers were strings, the format string would fail to parse, resulting in a crash.

The first line below avoids the unwieldiness of a match expression, taking advantage of the fact that the values of `in3D` and `extrudeAll` are uniform for all buildings. The second line uses the strongly typed initializer for match expressions to avoid string concatenation.

```suggestion
        let highlightedBuildingHeightExpression = NSExpression(format: "TERNARY(%@ AND (%@ OR $featureIdentifier IN %@), height, 0)", in3D, extrudeAll, buildings.map { $0.identifier })
        let colorsByBuilding = Dictionary(buildings.map { (NSExpression(forConstantValue: $0.identifier), NSExpression(forConstantValue: $0.highlightColor)) }) { (_, last) in last }
        let highlightedBuildingColorExpression = NSExpression(forMGLMatchingKey: NSExpression(forVariable: "featureIdentifier"), in: colorsByBuilding, default: NSExpression(forConstantValue: buildingDefaultColor))
```

Reply:
I hadn’t tried out these expressions before suggesting them. This is probably closer to running without issues:

```swift
NSExpression(format: "TERNARY(%@ = TRUE AND (%@ = TRUE OR $featureIdentifier IN %@), height, 0)", in3D as NSValue, extrudeAll as NSValue, buildings.map { $0.identifier })
```

Reply:
This one works, thanks.